src/metabase/driver/sql_jdbc/connection.clj

```diff
@@ -137,10 +137,35 @@ For setting the maximum, see [MB_APPLICATION_DB_MAX_CONNECTION_POOL_SIZE](#mb_ap
-   ;; kill connections after this amount of time if they haven't been returned -- this should be the same as the query
-   ;; timeout. This theoretically shouldn't happen since the QP should kill things after a certain timeout but it's
-   ;; better to be safe than sorry -- it seems like in practice some connections disappear into the ether
+   ;; [From dox] Seconds. If set, if an application checks out but then fails to check-in [i.e. close()] a Connection
+   ;; within the specified period of time, the pool will unceremoniously destroy() the Connection. This permits
+   ;; applications with occasional Connection leaks to survive, rather than eventually exhausting the Connection
+   ;; pool. And that's a shame. Zero means no timeout, applications are expected to close() their own
+   ;; Connections. Obviously, if a non-zero value is set, it should be to a value longer than any Connection should
+   ;; reasonably be checked-out. Otherwise, the pool will occasionally kill Connections in active use, which is bad.
+   ;;
+   ;; This should be the same as the query timeout. This theoretically shouldn't happen since the QP should kill
+   ;; things after a certain timeout but it's better to be safe than sorry -- it seems like in practice some
+   ;; connections disappear into the ether
    "unreturnedConnectionTimeout"  (jdbc-data-warehouse-unreturned-connection-timeout-seconds)
+   ;; [From dox] If true, and if unreturnedConnectionTimeout is set to a positive value, then the pool will capture
+   ;; the stack trace (via an Exception) of all Connection checkouts, and the stack traces will be printed when
+   ;; unreturned checked-out Connections timeout. This is intended to debug applications with Connection leaks, that
+   ;; is applications that occasionally fail to return Connections, leading to pool growth, and eventually
+   ;; exhaustion (when the pool hits maxPoolSize with all Connections checked-out and lost). This parameter should
+   ;; only be set while debugging, as capturing the stack trace will slow down every Connection check-out.
+   ;;
+   ;; Although we set this unconditionally to true here, note that these exceptions are not actually logged by
+   ;; default because we set the com.mchange log level to ERROR in our log4j2.xml config, and c3p0 logs the exceptions
+   ;; at INFO level. Therefore, you need to update the log level to INFO via a custom log4j config in order to see the
+   ;; stack traces in the logs.
+   ;;
+   ;; As noted in the C3P0 docs, this does add some overhead to create the Exception at Connection checkout.
+   ;; criterium/quick-bench indicates this is ~600ns of overhead per Exception created on my laptop. This doesn't seem
+   ;; worth adding another user-configurable knob to disable it given that we also unconditionally set
+   ;; testConnectionOnCheckout, above, which adds an order of magnitude more overhead per checkout (but still deemed
+   ;; acceptable).
+   "debugUnreturnedConnectionStackTraces" true
```

Comment:
Alternatively, we could

1. Make this user-configurable via a `defsetting`
2. Only enable if the `com.mchange` log4j level is `INFO` or higher

Reply:
I wound up putting this behind a `defsetting` that defaults to false and logging a warning if the setting is enabled and the C3P0 log level hasn't been raised.